File: smart-wheelchair/api/typescript/src/sensors/camera.ts
/**
 * @file camera.ts
 * @description WIA Smart Wheelchair Camera Interface
 * @version 1.0.0
 */

import {
  Header,
  Vector3D,
  PointCloud,
  BaseSensorConfig,
  SensorStatus,
  SensorCallback,
  Disposable,
} from './types';

/**
 * Camera type
 */
export type CameraType = 'rgb' | 'depth' | 'rgbd' | 'stereo' | 'thermal';

/**
 * Image format
 */
export type ImageFormat =
  | 'rgb8'
  | 'bgr8'
  | 'rgba8'
  | 'bgra8'
  | 'mono8'
  | 'mono16'
  | 'depth16'
  | 'depth32f';

/**
 * Camera configuration
 */
export interface CameraConfig extends BaseSensorConfig {
  type: CameraType;
  resolution: Resolution;
  fps: number;
  fov: FOV;
  format: ImageFormat;
  autoExposure: boolean;
  autoWhiteBalance: boolean;
}

/**
 * Resolution
 */
export interface Resolution {
  width: number;
  height: number;
}

/**
 * Field of View
 */
export interface FOV {
  horizontal: number;       // degrees
  vertical: number;         // degrees
  diagonal?: number;        // degrees
}

/**
 * Camera intrinsic parameters
 */
export interface CameraIntrinsics {
  fx: number;               // Focal length x (pixels)
  fy: number;               // Focal length y (pixels)
  cx: number;               // Principal point x (pixels)
  cy: number;               // Principal point y (pixels)
  k1?: number;              // Radial distortion k1
  k2?: number;              // Radial distortion k2
  k3?: number;              // Radial distortion k3
  p1?: number;              // Tangential distortion p1
  p2?: number;              // Tangential distortion p2
}

/**
 * Camera frame (RGB image)
 */
export interface CameraFrame {
  header: Header;
  format: ImageFormat;
  width: number;
  height: number;
  step: number;             // bytes per row
  data: Uint8Array;
  encoding?: string;
}

/**
 * Depth image
 */
export interface DepthImage {
  header: Header;
  width: number;
  height: number;
  step: number;
  data: Uint16Array | Float32Array;  // mm (uint16) or meters (float32)
  minDepth: number;         // meters
  maxDepth: number;         // meters
  unit: 'mm' | 'meters';
}

/**
 * RGBD frame (combined)
 */
export interface RGBDFrame {
  header: Header;
  color: CameraFrame;
  depth: DepthImage;
  aligned: boolean;         // Whether depth is aligned to color
}

/**
 * Stereo image pair
 */
export interface StereoFrame {
  header: Header;
  left: CameraFrame;
  right: CameraFrame;
  baseline: number;         // meters, distance between cameras
}

/**
 * Camera presets
 */
export const CAMERA_PRESETS: Record<string, Partial<CameraConfig>> = {
  realsense_d435: {
    type: 'rgbd',
    resolution: { width: 640, height: 480 },
    fps: 30,
    fov: { horizontal: 87, vertical: 58 },
  },
  realsense_d455: {
    type: 'rgbd',
    resolution: { width: 1280, height: 720 },
    fps: 30,
    fov: { horizontal: 87, vertical: 58 },
  },
  oak_d: {
    type: 'rgbd',
    resolution: { width: 1280, height: 720 },
    fps: 30,
    fov: { horizontal: 71.86, vertical: 56.74 },
  },
  zed2: {
    type: 'stereo',
    resolution: { width: 1280, height: 720 },
    fps: 30,
    fov: { horizontal: 110, vertical: 70 },
  },
  usb_webcam: {
    type: 'rgb',
    resolution: { width: 640, height: 480 },
    fps: 30,
    fov: { horizontal: 60, vertical: 45 },
  },
};

/**
 * Camera Interface
 */
export interface CameraInterface extends Disposable {
  /** Current configuration */
  readonly config: CameraConfig;

  /** Current status */
  readonly status: SensorStatus;

  /** Camera intrinsics */
  readonly intrinsics: CameraIntrinsics;

  /** Initialize camera */
  initialize(): Promise<void>;

  /** Start streaming */
  start(): Promise<void>;

  /** Stop streaming */
  stop(): Promise<void>;

  /** Register frame callback */
  onFrame(callback: SensorCallback<CameraFrame>): void;

  /** Remove frame callback */
  offFrame(callback: SensorCallback<CameraFrame>): void;

  /** Get latest frame */
  getLatestFrame(): CameraFrame | null;

  /** Capture single frame */
  capture(): Promise<CameraFrame>;

  /** Set exposure (if supported) */
  setExposure?(value: number): Promise<void>;

  /** Set gain (if supported) */
  setGain?(value: number): Promise<void>;
}

/**
 * Depth Camera Interface (extends CameraInterface)
 */
export interface DepthCameraInterface extends CameraInterface {
  /** Get depth image */
  getDepthImage(): DepthImage | null;

  /** Register depth callback */
  onDepth(callback: SensorCallback<DepthImage>): void;

  /** Remove depth callback */
  offDepth(callback: SensorCallback<DepthImage>): void;

  /** Convert depth to point cloud */
  toPointCloud(depth: DepthImage, color?: CameraFrame): PointCloud;

  /** Set depth range */
  setDepthRange(min: number, max: number): Promise<void>;
}

/**
 * RGBD Camera Interface
 */
export interface RGBDCameraInterface extends DepthCameraInterface {
  /** Get aligned RGBD frame */
  getRGBDFrame(): RGBDFrame | null;

  /** Register RGBD callback */
  onRGBD(callback: SensorCallback<RGBDFrame>): void;

  /** Remove RGBD callback */
  offRGBD(callback: SensorCallback<RGBDFrame>): void;

  /** Enable/disable depth-color alignment */
  setAlignment(enabled: boolean): Promise<void>;
}

/**
 * Convert depth image to point cloud
 */
export function depthToPointCloud(
  depth: DepthImage,
  intrinsics: CameraIntrinsics,
  color?: CameraFrame
): PointCloud {
  const points: Vector3D[] = [];
  const colors: { r: number; g: number; b: number }[] = [];

  const depthScale = depth.unit === 'mm' ? 0.001 : 1.0;

  for (let v = 0; v < depth.height; v++) {
    for (let u = 0; u < depth.width; u++) {
      const idx = v * depth.width + u;
      const d = (depth.data[idx] as number) * depthScale;

      if (d <= 0 || d < depth.minDepth || d > depth.maxDepth) {
        continue;
      }

      // Deproject pixel to 3D point
      const x = (u - intrinsics.cx) * d / intrinsics.fx;
      const y = (v - intrinsics.cy) * d / intrinsics.fy;
      const z = d;

      points.push({ x, y, z });

      // Add color if available
      if (color && color.format === 'rgb8') {
        const colorIdx = (v * color.width + u) * 3;
        colors.push({
          r: color.data[colorIdx],
          g: color.data[colorIdx + 1],
          b: color.data[colorIdx + 2],
        });
      }
    }
  }

  return {
    header: depth.header,
    points,
    colors: colors.length > 0 ? colors : undefined,
  };
}

/**
 * Project 3D point to pixel coordinates
 */
export function projectPoint(
  point: Vector3D,
  intrinsics: CameraIntrinsics
): { u: number; v: number } | null {
  if (point.z <= 0) {
    return null;
  }

  const u = (point.x * intrinsics.fx / point.z) + intrinsics.cx;
  const v = (point.y * intrinsics.fy / point.z) + intrinsics.cy;

  return { u, v };
}

/**
 * Deproject pixel to 3D ray
 */
export function deprojectPixel(
  u: number,
  v: number,
  intrinsics: CameraIntrinsics
): Vector3D {
  return {
    x: (u - intrinsics.cx) / intrinsics.fx,
    y: (v - intrinsics.cy) / intrinsics.fy,
    z: 1.0,
  };
}

/**
 * Apply undistortion to pixel coordinates
 */
export function undistortPoint(
  u: number,
  v: number,
  intrinsics: CameraIntrinsics
): { u: number; v: number } {
  const k1 = intrinsics.k1 ?? 0;
  const k2 = intrinsics.k2 ?? 0;
  const k3 = intrinsics.k3 ?? 0;
  const p1 = intrinsics.p1 ?? 0;
  const p2 = intrinsics.p2 ?? 0;

  // Normalize coordinates
  const x = (u - intrinsics.cx) / intrinsics.fx;
  const y = (v - intrinsics.cy) / intrinsics.fy;

  const r2 = x * x + y * y;
  const r4 = r2 * r2;
  const r6 = r4 * r2;

  // Radial distortion
  const radial = 1 + k1 * r2 + k2 * r4 + k3 * r6;

  // Tangential distortion
  const xd = x * radial + 2 * p1 * x * y + p2 * (r2 + 2 * x * x);
  const yd = y * radial + p1 * (r2 + 2 * y * y) + 2 * p2 * x * y;

  // Back to pixel coordinates
  return {
    u: xd * intrinsics.fx + intrinsics.cx,
    v: yd * intrinsics.fy + intrinsics.cy,
  };
}

/**
 * Resize image (simple nearest neighbor)
 */
export function resizeImage(
  frame: CameraFrame,
  newWidth: number,
  newHeight: number
): CameraFrame {
  const channels = frame.format.includes('rgba') || frame.format.includes('bgra') ? 4 :
                   frame.format.includes('rgb') || frame.format.includes('bgr') ? 3 : 1;

  const newData = new Uint8Array(newWidth * newHeight * channels);
  const xRatio = frame.width / newWidth;
  const yRatio = frame.height / newHeight;

  for (let y = 0; y < newHeight; y++) {
    for (let x = 0; x < newWidth; x++) {
      const srcX = Math.floor(x * xRatio);
      const srcY = Math.floor(y * yRatio);
      const srcIdx = (srcY * frame.width + srcX) * channels;
      const dstIdx = (y * newWidth + x) * channels;

      for (let c = 0; c < channels; c++) {
        newData[dstIdx + c] = frame.data[srcIdx + c];
      }
    }
  }

  return {
    ...frame,
    width: newWidth,
    height: newHeight,
    step: newWidth * channels,
    data: newData,
  };
}

File: smart-wheelchair/api/rust/src/sensors/camera.rs
//! Camera sensor interface for WIA Smart Wheelchair

use super::types::*;

/// Camera type
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CameraType {
    Rgb,
    Depth,
    Rgbd,
    Stereo,
    Thermal,
}

impl Default for CameraType {
    fn default() -> Self {
        Self::Rgb
    }
}

/// Image format
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ImageFormat {
    Rgb8,
    Bgr8,
    Rgba8,
    Bgra8,
    Mono8,
    Mono16,
    Depth16,
    Depth32f,
}

impl Default for ImageFormat {
    fn default() -> Self {
        Self::Rgb8
    }
}

impl ImageFormat {
    /// Get bytes per pixel
    pub fn bytes_per_pixel(&self) -> usize {
        match self {
            Self::Mono8 => 1,
            Self::Mono16 | Self::Depth16 => 2,
            Self::Rgb8 | Self::Bgr8 => 3,
            Self::Rgba8 | Self::Bgra8 | Self::Depth32f => 4,
        }
    }
}

/// Camera resolution
#[derive(Debug, Clone, Copy, Default)]
pub struct Resolution {
    pub width: u32,
    pub height: u32,
}

impl Resolution {
    pub fn new(width: u32, height: u32) -> Self {
        Self { width, height }
    }

    pub fn pixel_count(&self) -> usize {
        (self.width * self.height) as usize
    }
}

/// Field of view
#[derive(Debug, Clone, Copy, Default)]
pub struct FieldOfView {
    pub horizontal: f32,  // degrees
    pub vertical: f32,    // degrees
}

/// Camera intrinsic parameters
#[derive(Debug, Clone, Copy)]
pub struct CameraIntrinsics {
    pub fx: f64,  // Focal length x (pixels)
    pub fy: f64,  // Focal length y (pixels)
    pub cx: f64,  // Principal point x (pixels)
    pub cy: f64,  // Principal point y (pixels)
    pub k1: f64,  // Radial distortion
    pub k2: f64,
    pub k3: f64,
    pub p1: f64,  // Tangential distortion
    pub p2: f64,
}

impl Default for CameraIntrinsics {
    fn default() -> Self {
        Self {
            fx: 500.0,
            fy: 500.0,
            cx: 320.0,
            cy: 240.0,
            k1: 0.0,
            k2: 0.0,
            k3: 0.0,
            p1: 0.0,
            p2: 0.0,
        }
    }
}

impl CameraIntrinsics {
    /// Project 3D point to pixel coordinates
    pub fn project(&self, point: &Vector3D) -> Option<(f64, f64)> {
        if point.z <= 0.0 {
            return None;
        }

        let u = (point.x * self.fx / point.z) + self.cx;
        let v = (point.y * self.fy / point.z) + self.cy;

        Some((u, v))
    }

    /// Deproject pixel to 3D ray (unit vector)
    pub fn deproject(&self, u: f64, v: f64) -> Vector3D {
        Vector3D {
            x: (u - self.cx) / self.fx,
            y: (v - self.cy) / self.fy,
            z: 1.0,
        }
    }

    /// Deproject pixel with depth to 3D point
    pub fn deproject_with_depth(&self, u: f64, v: f64, depth: f64) -> Vector3D {
        Vector3D {
            x: (u - self.cx) * depth / self.fx,
            y: (v - self.cy) * depth / self.fy,
            z: depth,
        }
    }
}

/// Camera configuration
#[derive(Debug, Clone)]
pub struct CameraConfig {
    pub camera_type: CameraType,
    pub resolution: Resolution,
    pub fps: u32,
    pub fov: FieldOfView,
    pub format: ImageFormat,
    pub frame_id: String,
    pub auto_exposure: bool,
    pub auto_white_balance: bool,
}

impl Default for CameraConfig {
    fn default() -> Self {
        Self {
            camera_type: CameraType::Rgb,
            resolution: Resolution::new(640, 480),
            fps: 30,
            fov: FieldOfView { horizontal: 60.0, vertical: 45.0 },
            format: ImageFormat::Rgb8,
            frame_id: "camera_link".to_string(),
            auto_exposure: true,
            auto_white_balance: true,
        }
    }
}

/// Camera frame (image)
#[derive(Debug, Clone)]
pub struct CameraFrame {
    pub header: Header,
    pub format: ImageFormat,
    pub width: u32,
    pub height: u32,
    pub step: u32,  // bytes per row
    pub data: Vec<u8>,
}

impl CameraFrame {
    pub fn new(header: Header, width: u32, height: u32, format: ImageFormat) -> Self {
        let bpp = format.bytes_per_pixel();
        let step = width * bpp as u32;
        let data = vec![0u8; (height * step) as usize];

        Self {
            header,
            format,
            width,
            height,
            step,
            data,
        }
    }

    /// Get pixel value at (x, y)
    pub fn get_pixel(&self, x: u32, y: u32) -> Option<&[u8]> {
        if x >= self.width || y >= self.height {
            return None;
        }

        let bpp = self.format.bytes_per_pixel();
        let offset = (y * self.step + x * bpp as u32) as usize;
        Some(&self.data[offset..offset + bpp])
    }
}

/// Depth image
#[derive(Debug, Clone)]
pub struct DepthImage {
    pub header: Header,
    pub width: u32,
    pub height: u32,
    pub data: Vec<u16>,       // millimeters
    pub min_depth: f32,       // meters
    pub max_depth: f32,       // meters
}

impl DepthImage {
    pub fn new(header: Header, width: u32, height: u32) -> Self {
        Self {
            header,
            width,
            height,
            data: vec![0u16; (width * height) as usize],
            min_depth: 0.1,
            max_depth: 10.0,
        }
    }

    /// Get depth at pixel (x, y) in meters
    pub fn get_depth(&self, x: u32, y: u32) -> Option<f32> {
        if x >= self.width || y >= self.height {
            return None;
        }

        let idx = (y * self.width + x) as usize;
        let depth_mm = self.data[idx];

        if depth_mm == 0 {
            None
        } else {
            Some(depth_mm as f32 / 1000.0)
        }
    }

    /// Convert to point cloud
    pub fn to_point_cloud(&self, intrinsics: &CameraIntrinsics) -> Vec<Vector3D> {
        let mut points = Vec::new();

        for v in 0..self.height {
            for u in 0..self.width {
                if let Some(depth) = self.get_depth(u, v) {
                    if depth >= self.min_depth && depth <= self.max_depth {
                        let point = intrinsics.deproject_with_depth(
                            u as f64,
                            v as f64,
                            depth as f64,
                        );
                        points.push(point);
                    }
                }
            }
        }

        points
    }
}

/// RGBD frame (combined color and depth)
#[derive(Debug, Clone)]
pub struct RgbdFrame {
    pub header: Header,
    pub color: CameraFrame,
    pub depth: DepthImage,
    pub aligned: bool,
}

/// Stereo image pair
#[derive(Debug, Clone)]
pub struct StereoFrame {
    pub header: Header,
    pub left: CameraFrame,
    pub right: CameraFrame,
    pub baseline: f32,  // meters
}

/// Camera interface trait
pub trait CameraInterface: Send + Sync {
    /// Get current configuration
    fn config(&self) -> &CameraConfig;

    /// Get current status
    fn status(&self) -> SensorStatus;

    /// Get camera intrinsics
    fn intrinsics(&self) -> &CameraIntrinsics;

    /// Initialize camera
    fn initialize(&mut self) -> Result<(), String>;

    /// Start streaming
    fn start(&mut self) -> Result<(), String>;

    /// Stop streaming
    fn stop(&mut self) -> Result<(), String>;

    /// Get latest frame
    fn get_latest_frame(&self) -> Option<CameraFrame>;

    /// Capture single frame
    fn capture(&mut self) -> Result<CameraFrame, String>;
}

/// Depth camera interface trait
pub trait DepthCameraInterface: CameraInterface {
    /// Get depth image
    fn get_depth_image(&self) -> Option<DepthImage>;

    /// Convert depth to point cloud
    fn to_point_cloud(&self, depth: &DepthImage) -> Vec<Vector3D>;

    /// Set depth range
    fn set_depth_range(&mut self, min: f32, max: f32) -> Result<(), String>;
}

/// RGBD camera interface trait
pub trait RgbdCameraInterface: DepthCameraInterface {
    /// Get aligned RGBD frame
    fn get_rgbd_frame(&self) -> Option<RgbdFrame>;

    /// Enable/disable depth-color alignment
    fn set_alignment(&mut self, enabled: bool) -> Result<(), String>;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_camera_intrinsics_project() {
        let intrinsics = CameraIntrinsics {
            fx: 500.0,
            fy: 500.0,
            cx: 320.0,
            cy: 240.0,
            ..Default::default()
        };

        let point = Vector3D::new(0.0, 0.0, 1.0);
        let (u, v) = intrinsics.project(&point).unwrap();

        assert!((u - 320.0).abs() < 0.01);
        assert!((v - 240.0).abs() < 0.01);
    }

    #[test]
    fn test_camera_intrinsics_deproject() {
        let intrinsics = CameraIntrinsics {
            fx: 500.0,
            fy: 500.0,
            cx: 320.0,
            cy: 240.0,
            ..Default::default()
        };

        let point = intrinsics.deproject_with_depth(320.0, 240.0, 1.0);
        assert!((point.x).abs() < 0.01);
        assert!((point.y).abs() < 0.01);
        assert!((point.z - 1.0).abs() < 0.01);
    }
}
